blog/2024-11-18-relay-ops.md

```diff
@@ -0,0 +1,60 @@
+
+## Firehose Fan-out and Rainbow
+
+Firehose fan-out servers have a single upstream connection to a relay firehose, and re-broadcast that stream to multiple subscribing clients. They maintain a local backfill window (allowing re-connection with a cursor), but do not implement endpoints like `getRecord` or `getRepo`, only `com.atproto.sync.subscribeRepos`. They do not re-validate the event stream (eg, don't verify signatures). They maintain the exact sequence numbers of the upstream relay, meaning that they are interchangeable with the relay and with sibling fan-out servers, and can be placed behind a load-balancer.
+
+Today we are releasing Rainbow, a fan-out service implemented in Go. The [source code is available](https://github.com/bluesky-social/indigo/tree/main/cmd/rainbow) in the `indigo` git repository, with the same MIT/Apache licensing as our other open source projects.
+
+In the future, we may run additional Rainbow instances outside our primary data centers, similar to how we offer Jetstream instances today. This distributes the networking over more routers, and could improve connection reliability and throughput for subscribers in regions outside North America.
+
+Going forward, we recommend that software which consumes from firehoses (including atproto SDKs) support HTTP redirects for WebSocket connections. This enables "pooling" behavior where a single hostname could route clients to multiple distinct servers, without use of a load-balancer.
+
+## What Else?
+
+As a reminder, the Relay synchronization API is the same as the PDS synchronization API, including the firehose. Relays are essential to the functionality of the atproto network: they are an operational and developer convenience, allowing consuming services to skip keeping track of which PDS instances are active.
+
+A classic way to scale services like the firehose is *sharding*, where the stream is split in to multiple parallel streams. atproto has a natural sharding key (the account DID), which means related events can be consistently routed to the correct shard. This will allow near-indefinite scaling of the firehose event rate, assuming compute and network resources are available.
+
+Part of what makes Relays difficult to operate at scale is that they function as both a "relay" (rebroadcasting events from PDS instances) and a full-network mirror (storing all repo contents). In the current protocol, it is necessary to combine both functions to fully verify repository operations, especially deletion of events. We think that clever implementations could make this work less resource intensive (for example, but not storing records themselves, just CIDs). It is also possible to validate *most* aspects of the event stream without a full copy of the repo tree, and there could be a role in the network for "non-mirroring" relays.
```

Comment:
`(for example, but not storing records themselves, just CIDs)`
I think there might be a word or more missing here?

Reply:
fixed